### ISSUE TITLE
Persisting and hydrating nested/non-nested class objects-lists-maps

### ISSUE THREAD

**Issue by rafaeldsousa** [quarrant/mobx-persist-store#55]
As mentioned [here](https://github.com/quarrant/mobx-persist-store/issues/43#issuecomment-852723608). I'm having problems persisting lists of class objects-lists-maps (in a few cases those are nested with other class objects-lists-maps).

Referring back to the @pinqy520 [mobx-persist](https://github.com/pinqy520/mobx-persist) (RIP), I can see that this was a built in feature using [serializr](https://www.npmjs.com/package/serializr). 

@codeBelt pointed to https://github.com/quarrant/mobx-persist-store/issues/41#issuecomment-834755866 (Thanks). Which seems to have a way around this, by implementing a custom `getItem` to the StoreAdapter.

I believe this to be of common enough use for a feature to be implemented, if possible, in the module. 

I'll have a go on an implementation based on @pinqy520 module and will post updates here, but if this is something that might already be cooking in house, that would also be helpful.

**Comment by codeBelt** [comment]
Currently there is no new features in the works for anything. 

The current work around as mentioned above is to create a custom storage adapter that has `setItem`, `removeItem`, and `getItem` methods. This allows you to modify the data before it hydrates the store.

Sample:

```ts
makePersistable(this, {
  name: 'SampleStore',
  properties: ['someProperty'],
  storage: {
    setItem: (key, data) => window.localStorage.setItem(key, data),
    removeItem: (key) => window.localStorage.removeItem(key),
    getItem: (key) => {
      const data = JSON.parse(window.localStorage.getItem(key));

      return {
        ...data,
        someProperty: someLogicThatTransformsTheData(data.someProperty),
      };
    },
  },
});
```

Maybe one of these libraries will help:

- https://github.com/typestack/class-transformer
- https://github.com/codeBelt/sjs-base-model

@rafaeldsousa If you want to create a codesandbox.io or stackblitz.com example with this issue so we might also work on a solution. Please also post your solution in this thread so we can see what you come up with.

@quarrant Do you have any ideas on how this might be handle in the library? I think the custom storage adapter allow for the most flexibility but it is worth thinking about how the library might solve this.

**Comment by quarrant** [comment]
I don't think it is a good idea to add transformation logic inside the library. You can call plainToClass in `getItem` inside storage adapter.

**Comment by rafaeldsousa** [comment]
I've roughly implemented a solution using serializr and seems to work well. Based on the same approach as mobx-persist has.
The thing would be implementing serialisation to all persistent properties. But then for the ones that are supposed to map back to a class we'd need an object with type and the targetClass. My current solution, which works, relies on including an addition decorator to the props that we persist, I think it's an additional step that makes it easier to map nested classes. In mobx-persist the decorator would already trigger the persistence, so when declaring the persistent store we wouldn't need to pass the props. 

This is the barebones implementation I've worked out.  (Just couldn't reference my github on codesandbox so it currently doesn't run, but my local implementation is basically the same).
https://codesandbox.io/s/mobx-persist-store-mapping-t9epq

Branch
https://github.com/rafaeldsousa/mobx-persist-store/tree/experimental-propmapping

**Comment by quarrant** [comment]
we will not use serializr because `mobx-persist-store` is marketed as a lightweight persist implementation. Without using heavy transformations.

We need found other solution

**Comment by rafaeldsousa** [comment]
Completely understand your point, serializr would add quite a bit of weight to the library. 

My thinking here is that even though we have a lightweight persist store library, it lacks the ability needed for a little more than just plain properties usage as I've presented above, the data structure isn't too complicated. 

From the usage we've been having this data structure became quite common, but might not be your case so far so you can't related to what I'm saying. 

The module as is, is basic and us(or just me) wanting to have a little more complicated data structure has to implement custom code that isn't as simple as mapping things from the start and have to install some other module just as or more heavy than serializr (or not as simple), meaning we'll have to carry that weight anyways.

So we have a scenario where you have a limited lightweight library and then another, that to expand it further we'd be giving the lightweight title. 

I might be bias to the serializr usage as it worked just as I need and it wasn't too complicated to implement, but I'd be happy to discuss any other possible approaches that would keep the library light and still bring all the same features serializr provides.

**Comment by quarrant** [comment]
I think you can do it in storage adapter.

I don’t want add some custom transformations, because code will be hard and will have a magic.

**Comment by rafaeldsousa** [comment]
How would it work in the Storage Adapter, with the classes and subclass (nesting) ? Would we have to map all in the persist declaration ? Having to know what are the classes and subclasses of the nest at the parent level seems quite a bit of a hassle. 

Not sure about how hard the code would have to be, but can understand magic, as in things we don't know how it work and just happen.

**Comment by codeBelt** [comment]
@rafaeldsousa Looking at your example. Why does your sub classes have their own `makeAutoObservable` in them? I could be wrong but when you have your top level class call `makeAutoObservable` and then add items to an array become observed and not need their own `makeAutoObservable`.

I am working on an example showing how to use a custom storage adapter to re-create plain objects into class but I don't think it will cover your situation. More to come.

**Comment by rafaeldsousa** [comment]
@codeBelt that was just residual copy and paste that I haven't deleted. No actual need `makeAutoObservable` in the subclasses.

I think I'll sadly have to revert back to `mobx-persist` and use a fork (as it hasn't been maintained for quite a while and that's why I was on the look out for an alternative) in which I upgrade the dependencies. That one seems to be the only lib that actually covers my use case  and it's quite straightforward. I have quite a few apps at this stage that rely on subclasses for storage so it wouldn't be ideal having a custom store adapter and going through the process of exposing all subclasses to a parent :/

**Comment by codeBelt** [comment]
@rafaeldsousa looks like `mobx-persist` is best for your apps.

I put together an example how to convert the persisted data back into classes with the current version of `mobx-persist-store`.
https://codesandbox.io/s/mobx-persist-store-hydrate-classes-3wmkg?file=/src/stores/User.store.ts

I agree with @quarrant in not adding a built in transformation layer within the library. Maybe the library could facilitate  transforming data with the `properties` property. Currently you can only pass an array of strings of keys to `properties`.


`properties: ["someProperty", "user", "users"],`

But what if `properties` could take string and object values so you can customize what data you wanted to transform?

```
properties: [
  'someProperty',
  { key: 'user', convert: (data) => new Model(data) },
  { key: 'users', convert: (data) => data.map((item) => new Model(item)) },
],
```

Full Example:

```ts
export class UserStore {
  user: IUser | null = null;
  users: IUser[] = [];
  someProperty: number = 0;

  constructor() {
    makeAutoObservable(this);

    makePersistable(this, {
      name: 'UserStore',
      properties: [
        'someProperty',
        { key: 'user', convert: (data) => new Model(data) },
        { key: 'users', convert: (data) => data.map((item) => new Model(item)) },
      ],
    });
  }
}
```

This is just an idea and I haven't tried it yet.

**Comment by rafaeldsousa** [comment]
@codeBelt Yep, in this case I'll have to use `mobx-persist`, which I didn't want as the module has been unmaintained for a while, but that will have to do it for now.

In terms of what you've presented, I've had that at some point in my fork of `mobx-persist-store`, but I've ran into issues with the subclasses, as the persistence wasn't defined in the same class. I was doing something pretty similar to what you've proposed, if the prop of `properties` is a string, just process it normally but if it's an object then do something else. But as I've mentioned before the subclassing was preventing me from doing that. At that point I kinda understood I had to map all classes/types needed before hand to do the transformations based on the map.

But it's alright, `mobx-persist-store` is great and has it's usage and I'll certainly be using for simpler case, it just doesn't fit all scenarios in this project.
